### DISCUSSION TITLE
아이템 48. 스트림 병렬화는 주의해서 적용하라

### DISCUSSION THREAD

**Discussion by coalong** [Study-2-Effective-Java/Effective-Java#121] [3. 과제]
# 아이템 48. 스트림 병렬화는 주의해서 적용하라

## 아이템 설명
> 자바는 동시성 프로그래밍 측면에서 항상 신경 썼다.
그렇지만, 이를 올바르고 작성하는 것은 어렵다. 
이번 아이템은 스트림 병렬화의 문제점을 예로 설명한다.
> 

### 자바의 동시성 프로그래밍
- 처음 릴리즈부터 지원 : 스레드, 동기화, wait/notify 지원
- Java 5 : java.util.concurrent 라이브러리, 실행자(Executor) 프레임워크 지원
- Java 7 : 고성능 병렬 분해(parallel decom-position) 프레임워크인 [포크-조인(fork-join) 패키지](https://www.baeldung.com/java-fork-join)를 추가
- Java 8 : parallel 메서드를 한 번 호출하면 파이프라인을 병렬 실행할 수 있는 스트림 지원

> 이처럼 자바로 동시성 프로그램을 작성하기 쉬워지고는 있지만, **안정성(safety)과 응답 가능(liveness) 상태를 유지하기 위해 코드를 올바르게 작성하는 것은 어려운 작업**이다.

### 동시성(Concurrency) 
- 싱글 코어에서 멀티스레드를 동작시키기 위한 방식으로, 멀티 태스킹을 위해 **여러 개의 스레드가 번갈아가면서 실행되는 성질** 을 말한다.  (== 여러 작업이 동시에 실행되고 있는 것처럼 구현되는 것)

### 병렬성(Parallelism)
- **멀티 코어에서 멀티스레드를 동작** 시키는 방식으로, **한개 이상의 스레드를 포함하는 각 코어들이 동시에 실행되는 성질** 을 말한다. (== 여러 작업이 실제로 동시에 실행되고 있는 것)

![image](https://user-images.githubusercontent.com/71637466/220313829-6a0de9c8-83dc-46cd-9a3d-023f6ae14e43.png)

### 스트림을 사용해 처음 20개의 메르센 소수를 생성하는 프로그램
(아이템 45)
```java
public static void main(String[] args) {
        long startTime = System.currentTimeMillis();

        primes().map(p -> TWO.pow(p.intValueExact()).subtract(ONE))
                .filter(mersenne -> mersenne.isProbablePrime(50))
                .limit(20)
                .forEach(System.out::println);
        long endTime = System.currentTimeMillis();
        System.out.println(String.format("코드 실행 시간: %20dms", endTime - startTime));
    }

    static Stream<BigInteger> primes() {
        return Stream.iterate(TWO, BigInteger::nextProbablePrime);
    }
```

> 위 코드는 소수를 출력하는 코드이다. 
`코드 실행 시간:  17402ms`

> 스트림 파이프라인의 parallel() 을 호출하여 실행하면?
```java
public static void main(String[] args) {
        long startTime = System.currentTimeMillis();

        primes().map(p -> TWO.pow(p.intValueExact()).subtract(ONE))
                .parallel()
                .filter(mersenne -> mersenne.isProbablePrime(50))
                .limit(20)
                .forEach(System.out::println);
        long endTime = System.currentTimeMillis();
        System.out.println(String.format("코드 실행 시간: %20dms", endTime - startTime));
    }

    static Stream<BigInteger> primes() {
        return Stream.iterate(TWO, BigInteger::nextProbablePrime);
    }
```
`응답불가` + `CPU 많이 사용`
 

### 왜 이런 현상이 이러난 걸까?
- 스트림 라이브러리가 이 파이프라인을 병렬화하는 방법을 찾아내지 못했기 때문이다.
- **Stram.iterate 거나 중간 연산으로 limit 를 쓰면 파이프라인 병렬화로는 성능 개선을 기대할 수 없다.**
  - 위의 코드는 두 경우에 모두 해당된다.
- limit을 다룰 때 CPU 코어가 남는다면 원소를 몇 개 더 처리 후 제한된 개수 이후의 결과를 버려도 아무런 해가 없다고 가정한다. 
## 스트림 병렬화를 적용하는 경우
- 스트림 병렬화는 오직 성능 최적화 수단이다.
- 따라서 성능 테스트를 통해 병렬화를 적용할 가치가 있을 때만 사용한다.
- 스트림을 사용하기 좋은 조건
  - 스트림의 소스가 ArrayList, HashMap, HashSet, ConcurrentHashMap의 인스턴스나 배열, int, long 범위일 때 병렬화의 효과가 가장 좋다.
    1. 이 자료구조들은 모두 데이터를 원하는 크기로 정확하고 손쉽게 나눌수 있어 일을 다수의 스레드에 분배하기 좋다.
    2. 순차적으로 실행할 떄 참조 지역성이 뛰어나다.
       - 참조 지역성이 뛰어나다는 것은 메모리에 연속해서 저장되어 있다는 의미
       - 참조 지역성이 낮으면 스레드 작업에서 주메모리 -> 캐시 메모리로 전송되어 오는 시간이 길어진다.
       - 기본 타입의 배열이 참조 지역성이 뛰어나다.
- 조건이 잘 갖춰지면 **parallel 메서드 호출 하나로 거의 프로세서 코어 수에 비례하는 성능 향상**을 만끽할 수 있다.

## 핵심정리
- 계산도 올바로 수행하고 성능도 빨라질 거라는 확신 없이는 스트림 파이프라인 병렬화는 시도 하지 말라.
- 스트림을 잘못 병렬화하면 프로그램을 오동작하거나 성능을 급격히 떨어뜨린다.
- 스트림의 소스가 ArrayList, HashMap, HashSet, ConcurrentHashMap의 인스턴스나 배열, int, long 범위일 때 병렬화의 효과가 가장 좋다.
- 즉, 계산도 정확하고 성능도 좋아졌음이 확실해졌을 때만 스트림 병렬화를 사용하라.

**Comment by corock** [comment]
백엔드 개발에서 성능 향상이라는 주제는 시니어로 가기 위해 자주 언급되는 주제인 것 같습니다.
어려운 내용이었을텐데 잘 정리해주셔서 감사합니다~!
